addons/mail/static/src/messaging/component/chat_window_hidden_menu/chat_window_hidden_menu.js

```diff
@@ -100,8 +100,11 @@ class ChatWindowHiddenMenu extends Component {
      * @param {MouseEvent} ev
      */
     _onClickToggle(ev) {
-        ev.stopPropagation();
-        this.env.messaging.chatWindowManager.toggleHiddenMenu();
+        if (this.env.messaging.chatWindowManager.isHiddenMenuOpen) {
+            this.env.messaging.chatWindowManager.closeHiddenMenu();
+        } else {
+            this.env.messaging.chatWindowManager.openHiddenMenu();
+        }
```

Comment:
I know you don't like toggle, but I really think this is the way to go. This change is bad, because it goes against the idea that components should be as dumb as possible: just render and have entry-points of user interactions. Management of these user interactions should happen in entity world, because business logic is there. With change here, you make components too smart, which goes against less complexities in components.

Reply:
This is component logic. The only reason it is a toggle is because that's how we coded the interface.
If there was one button to show it and one button to hide it, each would have its own handler without condition.

Toggle in entity is totally unreliable because it depends on pre-existing state and we have potentially async code anywhere.

Reply:
`ChatWindowManager` is a UI entity, so makes sense that it handles toggle. The idea is to dispatch user interactions to entity as soon as possible, no intelligence in the processing of these user interactions in component.

I don't agree with stating that toggle is unreliable. It's normal that toggle depends on pre-existing state, that's the whole purpose of toggle, and "potentially async code everywhere" is not a genuine reason to avoid toggle. Async behaviour always add more complexity, and we came to many problems in components so that's the reason we put them in entity. We don't add async just for the sake of having async, so we still need to be cautious whenever we have async.

Reply:
The intent is to show a menu or to hide a menu depending on the situation, is it never to toggle it randomly.

With a real toggle, if two parts of the code call toggle at the same time, you would except to get back to original state.

If I click on a button to show something, and automated code from elsewhere also asks to show it, I want it to be shown at the end, not to be toggle back to original closed state.

Reply:
This is a programming error if you call toggle twice and in the end you get back to original state. I strongly believe this is a very bad reason to de-structure such abstractions (toggle => explicit open or close in component world): a programming error must be fixed, not add boilerplate code to silently hide it: unit tests are there to catch such technical issues that prevent features to work as expected.

Reply:
I think the main example I would use is the following:

- user click to open
- owl is taking too long to render for whatever reason (browser management of animation frame)
- user click to open again, thinking he previously misclicked
- render is happening, but menu is still closed due to double toggle, user has to click once again...

This should not happen, but I just realized that my solution is not preventing it either because it is doing the condition based on the entity state, and not based on the last state that was there at render...

Either there should be one open event and one click event (differently bound on template at render time), or there should be a save of the local state after render, in which cases it becomes obvious why this code must be in the component.

I will agree however, that as it is (which is using entity state to make the condition) it wouldn't matter to move the condition into the entity as you suggest.

Reply:
The problem you raise is not related to inconsistency of toggle, but a performance issue. We shouldn't use weird tricks like that as camouflage for shattered code.

Reply:
It is totally related to inconsistency of toggle, as it highlights the importance of using the appropriate "previous state".

When the action is coming from the user, the only valid "previous state" (in regard to toggle) is the one that is displayed to him at the time of the action. It is based on this current state and nothing else that he decided to click, and based on this current state that he is expecting a result.
"I visually see this menu as closed right now, so I'm clicking to open it."
There is nothing else to it, the user is not making is decision based on async entity state.

I'm now convinced that the only proper solution to do toggle with OWL is to save locally inside the component the last rendered state.

Reply:
Just no, entity reflects the logical state of the component. If the component displays closed state, user clicks to open, component still display closed due to being super slow for some reasons, user clicks again, and eventually toggle + toggle = original state, then that's a problem on potential slowness of OWL (e.g. misusage of animation frame), or badly coded async behaviour somewhere else like in entities. But from a logical aspect, user click on button so entity should be aware that there was a click for toggle. Maybe logic of entity could manage "drop" of concurrent toggle actions, so that eventually a single toggle is registered, but that depends on entity code and is manage in entity code specifically.

Component should never ever do special logic like that, they should simply dispatch to entity, nothing more.

Reply:
>  entity reflects the logical state of the component

Unfortunately it is not the case, entity reflects the desired logical state of components (in the future), but not the current state of components.
Whether you want it or not OWL is currently doing buffered async rendering and we have to work with it.

> But from a logical aspect, user click on button so entity should be aware that there was a click for toggle.

The user intent is **not** to toggle, it is to **open** if he sees it closed, or to **close** it if he sees it open.

Doing it any other way would imply:
"I don't care what is the current state, and I don't care what will be the final state, I just want you to change it."

This is never going to be the goal of the user with such visual elements.


=> Due to technical limitation of point 1 and functional reason of point 2, the ideal way that you suggest just is not possible currently.

Reply:
Note that this would also be solved with immutable/versioning, where the "toggle" information would be coupled with the properly "versioned" state.

We're just not there yet.